rust_src/src/fns.rs

```diff
@@ -598,4 +599,55 @@ pub fn copy_sequence(mut arg: LispObject) -> LispObject {
     }
 }
 
+/// Check that ARRAY can have a valid subarray [FROM..TO),
+/// given that its size is SIZE.
+/// If FROM is nil, use 0; if TO is nil, use SIZE.
+/// Count negative values backwards from the end.
+/// Set *IFROM and *ITO to the two indexes used.
+pub fn validate_subarray_rust(
+    array: LispObject,
+    from: LispObject,
+    to: LispObject,
+    size: isize,
+) -> (EmacsInt, EmacsInt) {
+    // convert from, to and size to Emacs Int
+    let int_size = EmacsInt::from(LispObject::from(size));
+    let mut int_from = from.map_or(0, EmacsInt::from);
+    let mut int_to = to.map_or(int_size, EmacsInt::from);
+
+    // make negative numbers reverse array
+    if int_from < 0 {
+        int_from += int_size;
+    }
+    if int_to < 0 {
+        int_to += int_size;
+    }
+
+    // check if from is less than to, or if from or to are out of range.
+    if !(0 <= int_from && int_from <= int_to && int_to <= int_size) {
+        unsafe {
+            args_out_of_range_3(array, LispObject::from(int_from), LispObject::from(int_to));
```

Comment:
Whereever the C code needed these fixed argument number macros remacs usually has a Rust macro. You can find them in rust_src/src/eval_macros.rs. In this case you want `args_out_of_range!()`.